sheryew iP

sheryew opened the pull request:
## DukePro
> "Your mind is for having ideas, not holding them." - David Allen [(Source)](https://dansilvestre.com/productivity-quotes)

DukePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,
1. download it from [here]().
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😄 

And it is **FREE**!

Features:
- [x] Managing Tasks
- [x] Managing deadlines
- [x] Searching Tasks
- [ ] GUI (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

sheryew committed:
Level 0.

Rename, Greet, Exit

sheryew committed:
Echo

Echo user commands

sheryew committed:
Level 2

Add the ability to store whatever text entered by the user and display them back

sheryew committed:
Level 3

Add the ability to mark tasks as done.

sheryew committed:
Level 4

ToDos, Events, Deadlines

sheryew committed:
A-TextUiTesting

Automated Text UI Testing
-- Removed the code for initialising new scanner from the while loop

sheryew committed:
Level 5

Handle Errors

sheryew committed:
Level 6

Delete
Also edited a minor bug I found in level 5

sheryew committed:
Level 7. Save

sheryew committed:
Level 8

sheryew committed:
A-MoreOOP

sheryew committed:
Delete random

sheryew committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

sheryew committed:
Merge remote-tracking branch 'upstream/add-gradle-support' into A_Packages

sheryew committed:
A-Packages

Removed README.md as it was conflicting with my migration of Duke into duke package repo

sheryew committed:
Merge branch 'A_Packages'

sheryew committed:
A-JUnit

sheryew committed:
A-Jar

sheryew committed:
Add JavaDoc comments

sheryew committed:
Tweak the code to comply with a coding standard

sheryew committed:
Find a task by searching for a keyword

sheryew committed:
level 10. GUI

p-xp commented on src/main/java/DialogBox.java:
```diff
@@ -0,0 +1,84 @@
+import javafx.scene.layout.HBox;
+
+/**
+ * An example of a custom control using FXML.
+ * This control represents a dialog box consisting of an ImageView to represent the speaker's face and a label
+ * containing text from the speaker.
```
Very clear comment!

p-xp commented on src/main/java/DialogBox.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    /**
+     * Generates a default DialogBox GUI whenever user's prompts DUKE
```
Typo in user's?

p-xp commented on src/main/java/DialogBox.java:
```diff
@@ -0,0 +1,84 @@
+     * Compared to the user's DialogBox, DUKE's DialogBox is flipped to showcase contrast in response.
+     *
+     * @param text String Object. DUKE's response.
+     * @param img Image Object. Duke's default picture.
```
Generally clear comment, try to keep styling of Duke (Duke vs DUKE) consistent for better readability.

p-xp commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,61 @@
+     * @return MMM-D-YYYY representation from LocalDate or LocalDateTime object
+     * @throws UnsupportedOperationException If object is not of LocalDate or LocalDateTime
+     */
+    public String generateNewDateString() {
+        if (this.by instanceof LocalDate) {
+            return ((LocalDate) this.by).format(DateTimeFormatter.ofPattern("MMM d yyyy"));
+        } else if (this.by instanceof LocalDateTime) {
+            return ((LocalDateTime) this.by).format(DateTimeFormatter.ofPattern("MMM d yyyy, ha"));
+        }
+        throw new UnsupportedOperationException("Unsupported Temporal type");
+    }
```
Clear logical flow!

p-xp commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,236 @@
+     * @return String generated from the UI class. Aids in generation of DialogBox.
+     * @throws Exception if the command is Bye
+     */
+    public String validateParser(TaskList tasks) throws Exception {
```
This method is quite long and violates the 30 LoC guideline ([source](https://nus-cs2103-ay2324s1.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-long-methods))

Perhaps you could consider moving the actions into separate methods of their own?

e.g:
```
if (Objects.equals(action, "todo")) {
   tryCreateTask();
}
```

p-xp reviewed (commented):
LGTM, just some nits.

wnchan commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,83 @@
+     * @throws DukeException If Deadline object is not instantiated properly
+     * @throws IOException If file is not located or not available
+     */
+    public ArrayList<Task> load(String filePath) throws DukeException, IOException {
```
Code looks good to me. However, it could be further improved by making Task an abstract class, so you can create methods for each case.

wnchan commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,41 @@
+    public ArrayList<Task> filter(String keyword) {
+        ArrayList<Task> filteredTask = new ArrayList<>();
+        for (int i = 0; i < userData.size(); i++) {
+            Task task = userData.get(i);
```
Nice filtering method!

wnchan commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,62 @@
+package duke;
+
+public class Ui {
```
LGTM!

wnchan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,236 @@
+            }
+        }
+        return true;
+    }
```
Could you combine these methods?

wnchan commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,85 @@
+    public String getResponse(String user_input) {
+        try {
+            Parser parser = new Parser(user_input);
+            String answer = parser.validateParser(tasks);
```
Nice flow!

wnchan reviewed (commented):
LGTM! Just some nits to fix.

sheryew committed:
Add assertions to Parser class

Parser Class: Validate and return appropriate responses from user's prompts

Parser Class allows invalid prompts from users (e.g. list a, find).

Include assertions in certain portions of the logic flow found in the Parser class.

Assertions prevent illogical prompts from affecting the operations of Duke.

sheryew committed:
Merge pull request #1 from sheryew/Assertions_Branch

Add assertions to Parser class

sheryew committed:
Improve Code Quality

sheryew committed:
Merge pull request #2 from sheryew/CodeQuality_Branch

Improve Code Quality

sheryew committed:
Update README.md

sheryew committed:
Provide a way to get reminders about task

Current Situation
Deadline Class do not provide a method to check if the deadline of a particular Task is within x days.

Why it needs to change
This allows users to make full use of the functionality of the DUKE application as well as the special LocalDate object in Deadline class.

What is being done about it
Deadline class has a method that checks if a particular Deadline Task's due date is within x days. Parser Class also has the appropriate conditional checks to ensure users prompts are correct

Why it is done that way
This ensures proper abstraction

sheryew committed:
Merge pull request #3 from sheryew/Extension_Branch

Provide a way to get reminders about task

sheryew committed:
Added Product Website

sheryew committed:
Update README.md

sheryew committed:
Update README.md

sheryew committed:
Changing build.gradle for JAR file

sheryew committed:
Merge branch 'master' of https://github.com/sheryew/ip
